src/test/scala/rise/autotuning.scala

```diff
@@ -40,10 +40,38 @@ class autotuning extends test_util.Tests {
   test("collect constraints") {
     val e: Expr = convolution
     autotune.collectConstraints(e, autotune.collectParameters(e)).foreach(println)
-    val constraints = collectConstraints(e, autotune.collectParameters(e))
+  }
+
+  test("substitute parameters") {
+    val e: Expr = convolution(32)
+    val constraints = autotune.collectConstraints(e, autotune.collectParameters(e))
+
+    val badParameters1 = Map(
+      NatIdentifier("vec", isExplicit = true) -> (5: Nat),
+      NatIdentifier("tile", isExplicit = true) -> (15: Nat)
+    )
+    assert(!autotune.checkConstraints(constraints, badParameters1))
+
+    val badParameters2 = Map(
+      NatIdentifier("vec", isExplicit = true) -> (4: Nat),
+      NatIdentifier("tile", isExplicit = true) -> (13: Nat)
+    )
+    assert(!autotune.checkConstraints(constraints, badParameters2))
+
+    /* FIXME: there is no `n >= tile` constraint collected
+    val badParameters3 = Map(
+      NatIdentifier("vec", isExplicit = true) -> (8: Nat),
+      NatIdentifier("tile", isExplicit = true) -> (64: Nat)
+    )
+    assert(!autotune.checkConstraints(constraints, badParameters3))
+    */
```

Comment:
we could add a tuning parameter `n` for the input size and give it a fixed value. Then we could use it in the constraints.

Reply:
We can already talk about `n` inside the constraints, and there are already constraints involving `n`. The issue is that the constraint collection does not collect any constraint of the form `n >= tile` although it is required for validity.